packages/go_router/test/go_route_information_parser_test.dart

```diff
@@ -0,0 +1,192 @@
+    expect(matches[1].extra, extra);
+    expect(matches[1].fullUriString, '/abc?def=ghi');
+    expect(matches[1].subloc, '/abc');
+    expect(matches[1].route, routes[0].routes[0]);
+  });
+
+  test('GoRouteInformationParser returns error when unknown route', () async {
+    final List<GoRoute> routes = <GoRoute>[
+      GoRoute(
+        path: '/',
+        builder: (_, __) => const Placeholder(),
+        routes: <GoRoute>[
+          GoRoute(
+            path: 'abc',
+            builder: (_, __) => const Placeholder(),
+          ),
+        ],
+      ),
+    ];
+    final GoRouteInformationParser parser = GoRouteInformationParser(
+      routes: routes,
+      redirectLimit: 100,
+      topRedirect: (_) => null,
+    );
+
+    final List<GoRouteMatch> matches = await parser
+        .parseRouteInformation(const RouteInformation(location: '/def'));
+    expect(matches.length, 1);
+    expect(matches[0].queryParams.isEmpty, isTrue);
+    expect(matches[0].extra, isNull);
+    expect(matches[0].fullUriString, '/def');
+    expect(matches[0].subloc, '/def');
+    expect(matches[0].error!.toString(),
+        'Exception: no routes for location: /def');
+  });
+
+  test('GoRouteInformationParser can work with route parameters', () async {
+    final List<GoRoute> routes = <GoRoute>[
+      GoRoute(
+        path: '/',
+        builder: (_, __) => const Placeholder(),
+        routes: <GoRoute>[
+          GoRoute(
+            path: ':uid/family/:fid',
+            builder: (_, __) => const Placeholder(),
+          ),
+        ],
+      ),
+    ];
+    final GoRouteInformationParser parser = GoRouteInformationParser(
+      routes: routes,
+      redirectLimit: 100,
+      topRedirect: (_) => null,
+    );
+
+    final List<GoRouteMatch> matches = await parser.parseRouteInformation(
+        const RouteInformation(location: '/123/family/456'));
+    expect(matches.length, 2);
+    expect(matches[0].queryParams.isEmpty, isTrue);
+    expect(matches[0].extra, isNull);
+    expect(matches[0].fullUriString, '/');
+    expect(matches[0].subloc, '/');
+
+    expect(matches[1].queryParams.isEmpty, isTrue);
+    expect(matches[1].extra, isNull);
+    expect(matches[1].fullUriString, '/123/family/456');
+    expect(matches[1].subloc, '/123/family/456');
+    expect(matches[1].encodedParams.length, 2);
+    expect(matches[1].encodedParams['uid'], '123');
+    expect(matches[1].encodedParams['fid'], '456');
+  });
+
+  test('GoRouteInformationParser can do top level redirect', () async {
+    final List<GoRoute> routes = <GoRoute>[
+      GoRoute(
+        path: '/',
+        builder: (_, __) => const Placeholder(),
+        routes: <GoRoute>[
+          GoRoute(
+            path: ':uid/family/:fid',
+            builder: (_, __) => const Placeholder(),
+          ),
+        ],
+      ),
+    ];
+    final GoRouteInformationParser parser = GoRouteInformationParser(
+      routes: routes,
+      redirectLimit: 100,
```

Comment:
Should there be a test for exhausting the redirect limit?

Reply:
There's one redirect limit test in `go_router_test.dart`, but it wouldn't hurt to test the parser too.

Reply:
added a test: [77c0ce9](https://github.com/flutter/packages/pull/2189/commits/77c0ce9665d3cad2b3f4953e3f1797d3ecec22be)